libbeat/processors/convert/convert.go

```diff
@@ -0,0 +1,366 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package convert
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+const logName = "processor.convert"
+
+func init() {
+	processors.RegisterPlugin("convert", New)
+}
+
+type processor struct {
+	config
+	log *logp.Logger
+
+	converted []interface{} // Temporary storage for converted values.
+}
+
+// New constructs a new convert processor.
+func New(cfg *common.Config) (processors.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, errors.Wrap(err, "fail to unpack the convert processor configuration")
+	}
+
+	return newConvert(c)
+}
+
+func newConvert(c config) (*processor, error) {
+	log := logp.NewLogger(logName)
+	if c.Tag != "" {
+		log = log.With("instance_id", c.Tag)
+	}
+
+	return &processor{config: c, log: log, converted: make([]interface{}, len(c.Fields))}, nil
+}
+
+func (p *processor) String() string {
+	json, _ := json.Marshal(p.config)
+	return "convert=" + string(json)
+}
+
+var ignoredFailure = struct{}{}
+
+func resetValues(s []interface{}) {
+	for i := range s {
+		s[i] = nil
+	}
+}
+
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	defer resetValues(p.converted)
+
+	// Validate the conversions.
+	for i, conv := range p.Fields {
+		v, err := convertField(event, conv.From, conv.Type)
+		if err != nil {
+			switch cause := errors.Cause(err); cause {
+			case common.ErrKeyNotFound:
+				if !p.IgnoreMissing {
+					return event, annotateError(p.Tag, errors.Errorf("field [%v] is missing, cannot be converted to type [%v]", conv.From, conv.Type))
```

Comment:
Should this still return an error when ignore_failure is set? I've seen other processors with both flags wouldn't

Reply:
I added a check for this.